### DISCUSSION TITLE
Why is my root directory and cwd always the same when using telescope keymaps and other keymaps?

### DISCUSSION THREAD

**Discussion by sdondley** [LazyVim/LazyVim#1293] [Q&A]
Whether I do `<leader>ff` or `<leader>fF`, they both open the telescope window at the current working directory. 

Similarly, `<leader>ft` and `<leader>fT` open a terminal in the cwd. 

I assume that the lowercase versions of the keymaps are supposed to open at the top level of the current git repo. I'm not sure why this is not working. Is there some config setting I'm missing?

**Comment by sdondley** [comment]
OK, found the issue. I was using a different copilot plugin, `"zbirenbaum/copilot.lua`. It attaches itself as an lsp. It was setting the root to the cwd and overriding the other lsp that was attached to the buffer. After disabling the plugin, everything worked as expected.

**Comment by sdondley** [comment]
OK, so I tried the official copilot plugin and ran into the same issue. Also tried the official config suggestion at https://www.lazyvim.org/extras/coding/copilot for copilot and same problem occurred.

So does turning on copilot break the ability to find the root directory of a project?

**Comment by sdondley** [comment]
OK, I think I may have found the source of the problem, but as I'm not very familiar with lua or telescope and LazyVim, I'm only guessing.

But from what I can tell, the problem lies in the get_root method in the `LazyVim/lua/lazyvim/util` file:

```
function M.get_root()
  ---@type string?
  local path = vim.api.nvim_buf_get_name(0)
  path = path ~= "" and vim.loop.fs_realpath(path) or nil
  ---@type string[]
  local roots = {}
  if path then
    for _, client in pairs(vim.lsp.get_active_clients({ bufnr = 0 })) do
      local workspace = client.config.workspace_folders
      local paths = workspace and vim.tbl_map(function(ws)
        return vim.uri_to_fname(ws.uri)
      end, workspace) or client.config.root_dir and { client.config.root_dir } or {}
      for _, p in ipairs(paths) do
        local r = vim.loop.fs_realpath(p)
        require("notify")(r)
        if path:find(r, 1, true) then
          roots[#roots + 1] = r
        end
      end
    end
  end
  table.sort(roots, function(a, b)
    return #a > #b                                     -- is this sort wrong? 
  end)
  ---@type string?
  local root = roots[1]
  if not root then
    path = path and vim.fs.dirname(path) or vim.loop.cwd()
    ---@type string?
    root = vim.fs.find(M.root_patterns, { path = path, upward = true })[1]
    root = root and vim.fs.dirname(root) or vim.loop.cwd()
  end
  ---@cast root string
  return root
end
```

It seems that when multiple paths are found, they are sorted using a sort function with this line: `return #a > #b`.

If I change this to: `return #a < #b`, everything works as expected.

**Reply by jyuan0** [reply]
> Whether I do <leader>ff or <leader>fF, they both open the telescope window at the current working directory.

I could see this happening if your CWD was also a LSP root dir, e.g. `cd ~/.config/nvim; nvim init.lua`.

I think it's doubtful the sort is wrong. As far as I can tell, this is the intended behavior. Judging from other issues and discussions about this, you should probably change the functionality as you wish in your local config. You could disable the functionality or remap the keys.

There are steps to disable the `get_root` functionality, or remap the keys in this answer: https://github.com/LazyVim/LazyVim/discussions/952

**Reply by sdondley** [reply]
> I could see this happening if your CWD was also a LSP root dir, e.g. cd ~/.config/nvim; nvim init.lua

Yeah, that's not the case. 

I dumped the paths in the for loop to try to see what was going on. The paths that were output were consistent with the "root" paths for each of the lsp's, one of them being copilot which always returned the cwd.

As I mentioned above, everything works as expected with no copilot plugins. As soon as I turn a copilot plugin on, that's when things break. 

So I'd say there's a good chance that sort is backwards. Because when I reverse the sort, everything works as expected again with copilot turned on.

**Reply by jyuan0** [reply]
> So I'd say there's a good chance that sort is backwards. Because when I reverse the sort, everything works as expected again with copilot turned on.

I guess you could submit a PR for the change and change it locally in the meantime. I haven't noticed this with the copilot plugin but maybe it's just my setup.

**Reply by sdondley** [reply]
With LazyVim installed, when you do `<leader>cl`, it will pop open a window showing you what each lsp reports for the root directory. When you are at the bottom of a tree in a project, what does yours show for copilot? Does it show the cwd?

**Reply by jyuan0** [reply]
Sorry, I'm not using Copilot at the moment, so I can't test this for you.

**Reply by sdondley** [reply]
Ok, so I took a little time and boned up on my lua and indeed, the "roots" table get sorted by the path (or root) with the longest string so the cwd will be longer than the root of the project and gets placed first in the the newly sorted "roots" table with the code above. A couple lines lower, that path (the cwd) is set to root and becomes that path searched by telescope:

`local root = roots[1]`

**Comment by jolars** [comment]
I am facing a related problem, but I think the problem is that the copilot LSP server reports the wrong root directory. It's always my home directory, and when it's the only LSP attached, the <root_dir> commands just search my home directory instead of the project directory. I've filed an issue with copilot.lua at https://github.com/zbirenbaum/copilot.lua/issues/212.
